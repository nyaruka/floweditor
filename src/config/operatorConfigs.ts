export interface Operator {
    type: string;
    verboseName: string;
    operands: number;
    // Default category name to use if specified
    categoryName?: string;
}

export interface OperatorMap {
    [propName: string]: Operator;
}

export const operatorConfigList: Operator[] = [
    { type: 'has_any_word', verboseName: 'has any of the words', operands: 1 },
    { type: 'has_all_words', verboseName: 'has all of the words', operands: 1 },
    { type: 'has_phrase', verboseName: 'has the phrase', operands: 1 },
    { type: 'has_only_phrase', verboseName: 'has only the phrase', operands: 1 },
    { type: 'has_beginning', verboseName: 'starts with', operands: 1 },
    {
        type: 'has_text',
        verboseName: 'has some text',
        operands: 0,
        categoryName: 'Has Text'
    },
    {
        type: 'has_number',
        verboseName: 'has a number',
        operands: 0,
        categoryName: 'Has Number'
    },
    {
        type: 'has_number_between',
        verboseName: 'has a number between',
        operands: 2
    },
    { type: 'has_number_lt', verboseName: 'has a number below', operands: 1 },
    {
        type: 'has_number_lte',
        verboseName: 'has a number at or below',
        operands: 1
    },
    { type: 'has_number_eq', verboseName: 'has a number equal to', operands: 1 },
    {
        type: 'has_number_gte',
        verboseName: 'has a number at or above',
        operands: 1
    },
    { type: 'has_number_gt', verboseName: 'has a number above', operands: 1 },
    { type: 'has_date', verboseName: 'has a date', operands: 0, categoryName: 'Has Date' },
    { type: 'has_date_lt', verboseName: 'has a date before', operands: 1 },
    { type: 'has_date_eq', verboseName: 'has a date equal to', operands: 1 },
    { type: 'has_date_gt', verboseName: 'has a date after', operands: 1 },
    { type: 'has_run_status', verboseName: 'has a run status of', operands: 1 },
    { type: 'has_group', verboseName: 'is in the group', operands: 1 },
    {
        type: 'has_phone',
        verboseName: 'has a phone number',
        operands: 0,
        categoryName: 'Has Phone'
    },
    {
        type: 'has_email',
        verboseName: 'has an email',
        operands: 0,
        categoryName: 'Has Email'
    },
    {
        type: 'has_error',
        verboseName: 'has an error',
        operands: 0,
        categoryName: 'Has Error'
    },
    {
        type: 'has_value',
        verboseName: 'is not empty',
        operands: 0,
        categoryName: 'Not Empty'
    }
];

export const operatorConfigMap: OperatorMap = operatorConfigList.reduce(
    (map: OperatorMap, operatorConfig: Operator) => {
        map[operatorConfig.type] = operatorConfig;
        return map;
    },
    {}
);

export type GetOperatorConfig = (type: string) => Operator;

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: string): Operator => operatorConfigMap[type];
