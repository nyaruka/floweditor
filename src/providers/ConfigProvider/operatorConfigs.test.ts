import { operatorConfigList, operatorConfigMap, getOperatorConfig } from './operatorConfigs';

describe('Providers: ConfigProvider: operatorConfigs', () => {
    it('should provide operator config', () => {
        expect(operatorConfigList).toMatchSnapshot();
        expect(operatorConfigMap).toMatchSnapshot();

        Object.keys(operatorConfigMap).forEach(key =>
            expect(getOperatorConfig(key)).toEqual(operatorConfigMap[key])
        );
    });
});
