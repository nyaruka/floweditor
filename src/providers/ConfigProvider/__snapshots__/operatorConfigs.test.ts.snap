// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Providers: ConfigProvider: operatorConfigs should provide operator config 1`] = `
Array [
  Object {
    "operands": 1,
    "type": "has_any_word",
    "verboseName": "has any of the words",
  },
  Object {
    "operands": 1,
    "type": "has_all_words",
    "verboseName": "has all of the words",
  },
  Object {
    "operands": 1,
    "type": "has_phrase",
    "verboseName": "has the phrase",
  },
  Object {
    "operands": 1,
    "type": "has_only_phrase",
    "verboseName": "has only the phrase",
  },
  Object {
    "operands": 1,
    "type": "has_beginning",
    "verboseName": "starts with",
  },
  Object {
    "operands": 2,
    "type": "has_number_between",
    "verboseName": "has a number between",
  },
  Object {
    "operands": 1,
    "type": "has_number_lt",
    "verboseName": "has a number below",
  },
  Object {
    "operands": 1,
    "type": "has_number_lte",
    "verboseName": "has a number at or below",
  },
  Object {
    "operands": 1,
    "type": "has_number_eq",
    "verboseName": "has a number equal to",
  },
  Object {
    "operands": 1,
    "type": "has_number_gte",
    "verboseName": "has a number at or above",
  },
  Object {
    "operands": 1,
    "type": "has_number_gt",
    "verboseName": "has a number above",
  },
  Object {
    "operands": 1,
    "type": "has_date_lt",
    "verboseName": "has a date before",
  },
  Object {
    "operands": 1,
    "type": "has_date_eq",
    "verboseName": "has a date equal to",
  },
  Object {
    "operands": 1,
    "type": "has_date_gt",
    "verboseName": "has a date after",
  },
  Object {
    "operands": 1,
    "type": "has_run_status",
    "verboseName": "has a run status of",
  },
  Object {
    "operands": 1,
    "type": "has_group",
    "verboseName": "is in the group",
  },
  Object {
    "categoryName": "Has Date",
    "operands": 0,
    "type": "has_date",
    "verboseName": "has a date",
  },
  Object {
    "categoryName": "Has Text",
    "operands": 0,
    "type": "has_text",
    "verboseName": "has some text",
  },
  Object {
    "categoryName": "Has Number",
    "operands": 0,
    "type": "has_number",
    "verboseName": "has a number",
  },
  Object {
    "categoryName": "Has Phone",
    "operands": 0,
    "type": "has_phone",
    "verboseName": "has a phone number",
  },
  Object {
    "categoryName": "Has Email",
    "operands": 0,
    "type": "has_email",
    "verboseName": "has an email",
  },
  Object {
    "categoryName": "Has Error",
    "operands": 0,
    "type": "has_error",
    "verboseName": "has an error",
  },
  Object {
    "categoryName": "Not Empty",
    "operands": 0,
    "type": "has_value",
    "verboseName": "is not empty",
  },
]
`;

exports[`Providers: ConfigProvider: operatorConfigs should provide operator config 2`] = `
Object {
  "has_all_words": Object {
    "operands": 1,
    "type": "has_all_words",
    "verboseName": "has all of the words",
  },
  "has_any_word": Object {
    "operands": 1,
    "type": "has_any_word",
    "verboseName": "has any of the words",
  },
  "has_beginning": Object {
    "operands": 1,
    "type": "has_beginning",
    "verboseName": "starts with",
  },
  "has_date": Object {
    "categoryName": "Has Date",
    "operands": 0,
    "type": "has_date",
    "verboseName": "has a date",
  },
  "has_date_eq": Object {
    "operands": 1,
    "type": "has_date_eq",
    "verboseName": "has a date equal to",
  },
  "has_date_gt": Object {
    "operands": 1,
    "type": "has_date_gt",
    "verboseName": "has a date after",
  },
  "has_date_lt": Object {
    "operands": 1,
    "type": "has_date_lt",
    "verboseName": "has a date before",
  },
  "has_email": Object {
    "categoryName": "Has Email",
    "operands": 0,
    "type": "has_email",
    "verboseName": "has an email",
  },
  "has_error": Object {
    "categoryName": "Has Error",
    "operands": 0,
    "type": "has_error",
    "verboseName": "has an error",
  },
  "has_group": Object {
    "operands": 1,
    "type": "has_group",
    "verboseName": "is in the group",
  },
  "has_number": Object {
    "categoryName": "Has Number",
    "operands": 0,
    "type": "has_number",
    "verboseName": "has a number",
  },
  "has_number_between": Object {
    "operands": 2,
    "type": "has_number_between",
    "verboseName": "has a number between",
  },
  "has_number_eq": Object {
    "operands": 1,
    "type": "has_number_eq",
    "verboseName": "has a number equal to",
  },
  "has_number_gt": Object {
    "operands": 1,
    "type": "has_number_gt",
    "verboseName": "has a number above",
  },
  "has_number_gte": Object {
    "operands": 1,
    "type": "has_number_gte",
    "verboseName": "has a number at or above",
  },
  "has_number_lt": Object {
    "operands": 1,
    "type": "has_number_lt",
    "verboseName": "has a number below",
  },
  "has_number_lte": Object {
    "operands": 1,
    "type": "has_number_lte",
    "verboseName": "has a number at or below",
  },
  "has_only_phrase": Object {
    "operands": 1,
    "type": "has_only_phrase",
    "verboseName": "has only the phrase",
  },
  "has_phone": Object {
    "categoryName": "Has Phone",
    "operands": 0,
    "type": "has_phone",
    "verboseName": "has a phone number",
  },
  "has_phrase": Object {
    "operands": 1,
    "type": "has_phrase",
    "verboseName": "has the phrase",
  },
  "has_run_status": Object {
    "operands": 1,
    "type": "has_run_status",
    "verboseName": "has a run status of",
  },
  "has_text": Object {
    "categoryName": "Has Text",
    "operands": 0,
    "type": "has_text",
    "verboseName": "has some text",
  },
  "has_value": Object {
    "categoryName": "Not Empty",
    "operands": 0,
    "type": "has_value",
    "verboseName": "is not empty",
  },
}
`;
